[2기-C 홍유석] JPA 게시판 피드백 반영

yshjft opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
JPA 게시판 구현 과제 입니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->

### 요구 사항 및 구현 여부
- [X] SpringDataJPA를 설정한다.
   - datasource: h2
- [X] 엔티티를 구성한다.
- [X] API를 구현한다.
    - [X] 게시글 조회
       - [X] 페이징 조회 (GET "/posts")
       - [X] 단건 조회 (GET "/posts/{id}")
    - [X] 게시글 작성 (POST "/posts")
    - [X] 게시글 수정 (POST "/posts/{id}")
- [X] REST-DOCS를 이용해서 문서화한다.

### api 상세
api에 대한 자세한 내용은 [REST-DOCS를 이용하여 문서화한 것](https://yshjft.github.io/jpa-board-api-docs/)으로 대체하겠습니다.

## ✅ 피드백 반영사항 
### 엔티티 수정
* JPA Auditing 적용하였습니다.

* **엔티티 기본 생성자의 접근 제한자 변경**
엔티티 기본 생성자의 접근 제한자를 protected로 변경하여 무분별한 객체 생성을 방지하도록 하였습니다.

* **엔티티 필드 수정**
  * ```@Column```을 이용하여 필드에 제약을 명시적으로 확인할 수 있도록 수정하였습니다.
  * nullalbe인 필드의 경우 wrapper 클래스를 사용하도록 수정하였습니다.

* **엔티티 생성자 수정**
엔티티의 생서자에서 id가 포함되지 않도록 수정하였습니다.

### 검증 추가
Service와 Entity에서 전달된 값을 검증하도록 수정하였습니다.

### 테스트 코드 수정
Controller 테스트의 결과가 DB에 영향을 받지 않도록 수정하였습니다.

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
* 부족한 부분 말씀해주시면 반영하여 수정하도록 하겠습니다!

yshjft committed:
feat: init project

yshjft committed:
feat: User, Post 엔티티 생성

yshjft committed:
feat: UserRepository와 이에 대한 test code 작성 완료

yshjft committed:
feat:  PostRepository와 이에 대한 테스트 코드 작성 완료

yshjft committed:
feat: 회원 가입 api 구현

yshjft committed:
feat: 예외 처리 구현

yshjft committed:
feat: 게시물 생성 api 구현 및 UserNotFound 예외 처리

yshjft committed:
fix: 게시물 생성 service layer 코드 수정

yshjft committed:
feat: 게시글 페이징 조회 구현

yshjft committed:
feat: 게시물 조회 구현

yshjft committed:
feat: 게시글 수정 구현

yshjft committed:
feat: 문서화 추가

yshjft committed:
fix: 엔티티 수정

(1) 기본 생성자의 접근 제한자를 protected로 변경

(2) 엔티티의 생성자에서 id 설정하지 않도록 수정

yshjft committed:
fix: 테스트 코드 수정 및 추가, dto와 엔티티 수정

(1) 테스트 코드 수정 및 추가
게시글 업데이트 테스트 추가

(2) dto와 엔티티 수정
age가 DB에서 nullable=true이므로 dto와 엔티티에서 해당 값의 타입을 int에서 Integer로 변경

yshjft committed:
fix: controller 테스트 코드 수정, dto 수정

(1) 테스트 코드 수정
controller 테스트 코드 DB에 영향을 받지 않도록 @SpringBootTest에서 @WebMvcTest로 수정

(2) dto 수정
- 페이징 정보를 담던 클래스의 이름을 MetaDataDto에서 PagingInfo로 변경
- 테스트 코드에서의 사용을 위해 일부 DTO에 equals and hashcode 추가

yshjft committed:
fix: jpa auditing 적용과 그에 따른 dto 변경

jpa auditing을 적용하였으며 ResultDto에서 createdAt과 updatedAt을 제거 하였습니다.

yshjft committed:
fix: getter 수정

사용하는 getter만 존재하도록 엔티티 클래스에서 @Getter를 제거하였습니다. 또한 각각의 엔티티에서 DTO를 반환할 수 있도록 메서드를 구현하였습니다.

yshjft committed:
fix: service와 entity에서 validation 추가

VictoryPark commented on src/main/java/com/prgrms/jpaboard/domain/post/controller/PostController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<ResponseDto> getPosts(@Validated @ModelAttribute PageParamDto pageParamDto) {
```
validated 어노테이션을 쓰셨는데, validation grouping 한 코드는 못찾겠네요 ㅠㅠ
validated 를 쓰신 이유가 있을까요?

yshjft replied:
데이터 바인딩시 제약조건을 검증하려고 사용하였는데 제가 ```@Valid```와 ```@Validated```의 차이를 제대로 모르고 지금까지 사용한거 같습니다. 이 부분에 대해 학습한후 수정하겠습니다.

VictoryPark commented on src/main/java/com/prgrms/jpaboard/global/common/domain/BaseEntity.java:
```diff
@@ -0,0 +1,38 @@
+    protected BaseEntity() {
+    }
+
+    public BaseEntity(String className, String createdBy) {
+        validateCreatedBy(className, createdBy);
+        this.createdBy = createdBy;
+    }
```
validate 로직을 만드신 것을 봤는데, 매번 className을 파라미터로 넣어서 로그를 찍는 action은 비즈니스 로직과 맞지 않다고 생각합니다.
logger를 활용하면 어느 클래스에서 exception이 발생했는지 알 수 있습니다.
그리고 AOP를 사용해도 className을 동적으로 받을 수 있다는 점 알려드립니당

VictoryPark reviewed (approved):
유석님!
코드가 많이 좋아지셨어요!
테스트 코드도 깔끔하고 단위 테스트 잘 구분해서 하셨네요 ㅎㅎ
자세한 사항은 코멘트 달아드렸으니, 확인해주세용 ㅎㅎ

HyeonseoJUNG commented on src/test/java/com/prgrms/jpaboard/domain/post/domain/PostTest.java:
```diff
@@ -0,0 +1,40 @@
+        post = Post.builder()
+                .title("test title")
+                .content("this is content of test")
+                .build();
```
post 생성할 때 createdBy 값이 안들어가서 테스트가 실패하는 것 같아요!!

HyeonseoJUNG commented on src/test/java/com/prgrms/jpaboard/domain/post/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,189 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@MockBean(JpaMetamodelMappingContext.class)
+@AutoConfigureRestDocs
+@WebMvcTest(PostController.class)
```
이번에 @WebMvcTest로 테스트 코드 수정하신 부분보고 많이 배웠습니다!!👍👍

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/post/domain/Post.java:
```diff
@@ -0,0 +1,125 @@
+                .build();
+    }
+
+    public void updateTitle(String title) {
+        this.title = title;
+        this.updatedAt = LocalDateTime.now();
+    }
+
+    public void updateContent(String content) {
+        this.content = content;
+        this.updatedAt = LocalDateTime.now();
+    }
```
BaseEntity에서 @LastModifiedDate을 통해 updatedAt 값을 넣어주는데
이 부분에서 LocalDateTime.now()로 또 넣어주는 이유가 있나요?

yshjft replied:
이거 제가 JpaAudtining 반영하고 수정하는걸 잊은거 같습니다;; 수정하겠습니다!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,101 @@
+        return ResultDto.updateResult(id);
+    }
+
+    private void validatePostCreateDto(PostCreateDto postCreateDto) {
+        String className = postCreateDto.getClass().getName();
+
+        validateUserId(className, postCreateDto.getUserId());
+        validateTitle(className, postCreateDto.getTitle());
+        validateContent(className, postCreateDto.getContent());
+    }
+
+    private void validatePostUpdateDto(PostUpdateDto postUpdateDto) {
+        String className = postUpdateDto.getClass().getName();
+
+        validateTitle(className, postUpdateDto.getTitle());
+        validateContent(className, postUpdateDto.getContent());
+    }
```
저는 DTO에 대한 validation은 DTO 안에서 해주는게 좋다고 생각합니다!
유석님은 어떻게 생각하시나요??

yshjft replied:
서비스 레이어에서 직접 DTO를 생성하여 사용하는 것이 아니고 생성된 DTO를 메서드에서 파라미터로 전달 받아 사용하기에 검증 로직을 DTO가 아닌 서비스 레이어에서 하게 되었습니다.

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,37 @@
+        return ResultDto.createResult(savedUser.getId());
+    }
+
+    private void validateUserRequestDto(UserRequestDto userRequestDto) {
+        String className = "UserRequestDto";
+
+        validateName(className, userRequestDto.getName());
+        validateAge(className, userRequestDto.getAge());
+        validateHobby(className, userRequestDto.getHobby());
+    }
```
이 부분도 마찬가지로 저는 DTO 안에서 validation해주는게 나은 것 같습니다!!

HyeonseoJUNG reviewed (commented):
유석님! 고생하셨습니다😄
이번에 리팩토링하신 부분도 보면서 많이 배웠습니다!👍

yshjft committed:
fix: 피드백 반영(1)

(1) createdBy 누락으로 인해 테스트를 실패하는 것을 수정

(2) jpaAutoAuditing을 사용하고 있으므로 update 메소드에서 updatedAt 수동 수정 제거

yshjft committed:
fix: 피드백 반영(1)

(1) createdBy 누락으로 인해 테스트를 실패하는 것을 수정

(2) jpaAutoAuditing을 사용하고 있으므로 update 메소드에서 updatedAt 수동 수정 제거

(3) 테스트 코드에서 변수명을 통해 의미를 파악할 수 있도록 수정

yshjft committed:
fix: 피드백 반영(2)

정확한 의미를 모르고 사용하였던 ```@Valid```와 ```@Validated```의 차이점에 대해 학습하고 이를 바탕으로 코드를 수정하였습니다. 현재 validatioin grouping을 사용하고 있지 않기에 ```@Validated```를 ```@Valid```로 변경하였습니다.
